### DISCUSSION TITLE
Using custom hook patterns from React?

### DISCUSSION THREAD

**Discussion by Rinaldo** [solidjs/solid#1140] [Q&A]
I just discovered this library and I really like the look of it so far. Coming from a react background though, I'm unsure if custom hooks to control components are recommended or if there are other patterns I should prefer.

React makes it easy to create 'controller' hooks where the return value of the hook controls a component. It almost looks like two way binding if you squint a little. Good use cases are things like `useInput`, `useForm`, or `useModal` where the hook returns a bundle of props for the component and those props can be read in the parent where the hook is defined.

An example from [react-hooks](https://react-hooks.org/docs/useInput)
```jsx
import { useInput } from "rooks";

const App = () => {
  const inputController = useInput("hello");

  return (
    <div>
      <input {...inputController} />
      <p>
        Value is <b>{inputController.value}</b>
      </p>
    </div>
  );
}
```

It seems like in solid you can do something similar but I'm not sure about passing reactive props to the hook
```jsx
import { mergeProps } from "solid-js";
import { createStore } from "solid-js/store";

const Input = (props) => (
  <label style={{ display: "block" }}>
    {props.label}
    <br />
    <input value={props.value} onInput={props.onInput} style={{ outline: props.valid ? null : "1px solid red" }} />
  </label>
)

const createInputController = ({
    initialValue = "",
    isValid = (_) => true
} = {}) => {
  const [store, setStore] = createStore({ value: initialValue, valid: isValid(initialValue) })
  const onInput = e => {
    const value = e.target.value
    setStore({ value: e.target.value, valid: isValid(value) })
  }

  return mergeProps({
    onInput
  }, store)
}

function App() {
  const inputController = createInputController()
  return (
    <div>
      <Input label="My Input" {...inputController} />
    </div>
  );
}
```

**Is the above pattern recommended?**

**What's the best way to pass reactive props to the `createInputController` function?**
_Maybe there could be a utility that would do the same normalization that's already done for components so if you're trying to pass a prop into a hook you don't have to manually wrap it in a getter. Or maybe the hook should be called within a createMemo? Would that be expensive?_

**Comment by atk** [comment]
You can just put onInput into your store, no need to wrap it or put the thing in a memo – react only needs that because its hooks are run on every render, whereas ours only run once.

**Reply by Rinaldo** [reply]
Ok so this is fine?
```jsx
const createInputController = ({ initialValue = "", isValid = (_) => true } = {}) => {
  const [store, setStore] = createStore({ value: initialValue, valid: isValid(initialValue) })
  const onInput = e => {
    const value = e.target.value
    setStore({ value: e.target.value, valid: isValid(value) })
  }
  setStore({ onInput })

  return store
}
```

But my bigger question was about passing reactive parameters into the hook. If component props were signals it would be easy to pass them, but since they're getters you have to turn them back into signals before passing them to keep reactivity. If you did that then inside the hook you'd have to remember that you're dealing with an object containing signals, not a props-like object containing getters. Basically the jsx transform is doing some work to normalize how components are passed props so you can pass a constant, a value from a store, a value from a signal, or a signal itself to a component and it all works the same. I think it would worthwhile to extend that functionality to hooks if they're going to be first class citizens in solid.

**Reply by atk** [reply]
A store or props getter will work as a reactive function inside an effect. In those cases you need an actual accessor, just use `() => props.x`.
